aggregator/src/aggregator/aggregation_job_creator.rs

```diff
@@ -568,166 +613,30 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
                 let vdaf = Arc::clone(&vdaf);
 
                 Box::pin(async move {
-                    // Find unaggregated client reports & existing unfilled batches.
-                    let (mut unaggregated_report_ids, outstanding_batches) = try_join!(
-                        tx.get_unaggregated_client_report_ids_for_task(task.id()),
-                        tx.get_outstanding_batches_for_task(task.id())
-                    )?;
-
-                    // First attempt to allocate unaggregated reports to existing unfilled batches,
-                    // then generate new batches as necessary. This iterator has no end and
-                    // therefore it is safe to unwrap the result of a call to `next`.
-                    let mut batch_iter = outstanding_batches
-                        .into_iter()
-                        .map(|outstanding_batch| (false, outstanding_batch))
-                        .chain(iter::repeat_with(|| {
-                            (
-                                true,
-                                OutstandingBatch::new(
-                                    *task.id(),
-                                    random(),
-                                    RangeInclusive::new(0, 0),
-                                ),
-                            )
-                        }));
+                    // Find unaggregated client reports.
+                    let unaggregated_report_ids = tx
+                        .get_unaggregated_client_report_ids_for_task(task.id())
```

Comment:
This will return some client reports, but the reports that are returned are returned independently of their client timestamps.

I think this would permit a failure mode where, even when there are more than enough client reports available to create aggregation jobs in the relevant batch interval(s), we just-so-happen to retrieve too few client reports per batch interval to create an aggregation job. (This might even be cascading: as we fail to create aggregation jobs, more reports will arrive living in more batch intervals, so we are even more likely to retrieve a set of client reports which don't permit creation of aggregation jobs.)

One non-solution would be to introduce an index on `(task_id, client_timestamp)` and use that to retrieve the oldest aggregatable reports. This doesn't necessarily fix the issue because we might end up in a situation where there are many old "straggler" reports which are returned each time, causing us to not see the newer reports.

One solution that _might_ work would be to introduce the above index, then use that to retrieve the _newest_ aggregatable reports. This will work better as long as incoming reports have roughly increasing client timestamps. But I suppose even here we might end up "stranding" some older reports behind a number of intervals without enough reports to create an aggregation job, which will then never be aggregated.

Maybe there's a better solution possible. WDYT?

Reply:
Yeah, fetching the newest unaggregated reports sounds promising. One other idea: we could solve this problem by aggregating straggler reports, turning our straggler report problem into a straggler batch problem (which we already have). We could adjust our aggregation job size heuristics, either based on report age, or adaptively based on the progress made by the aggregation job creator, and override the minimum batch size or minimum aggregation size. This would trade overhead on aggregating those stragglers for liveness and efficiency of the aggregation job creator.

On another note, time bucketing means we will have outstanding batches that never qualify for collection. (until they become expired, and get ignored and cleaned up) The database query that resolves `current_batch` requests will have to filter through these (and potentially run aggregations) to find completed batches. We could similarly add an `ORDER BY DESC` here to check newer batches first. This should ensure the query finds completed batches soon in the normal case, but it doesn't change that we'd have to run aggregations over all stragglers when we're caught up.

Reply:
I think we can start with "aggregate newest unaggregated reports" and go from there.

A heuristic might work, but getting the details right would be complicated. (also, nit: we can't override the minimum batch size without protocol text support, since it is also checked by the helper. we can of course override the minimum aggregation job size.)

Reply:
In that situation, we would only create outstanding batches and run the aggregation flow on them, but never run the collection flow, (assuming we never get enough late reports) so it shouldn't be a problem at the protocol level.

Reply:
I added the `ORDER BY` clause to the report query while rebasing; we already have an index covering those columns, `client_reports_task_and_timestamp_index`.